cmdstanpy/utils.py

```diff
@@ -1431,3 +1432,27 @@ def __enter__(self) -> Tuple[str, bool]:
     def __exit__(self, exc_type, exc_val, exc_tb) -> None:  # type: ignore
         if self._tmpdir:
             shutil.rmtree(self._tmpdir, ignore_errors=True)
+
+
+def write_stan_file(code: str, directory: Union[str, None] = None) -> str:
+    """
+    Write the model code to a file and return the path based on the code hash.
+
+    :param code: Code to write to a file.
+    :param directory: Directory to write the code to (defaults to the cmdstanpy
+        models directory).
+
+    :return: Path to the file containing the code..
+    """
+    # Generate the path based on the code and check whether it already exists.
+    directory = directory or os.path.join(cmdstan_path(), 'models')
+    digest = hashlib.sha1(code.encode()).hexdigest()
+    stan_file = os.path.join(directory, f'{digest}.stan')
+    if os.path.isfile(stan_file):
+        return stan_file
+
+    # Write the file to disk.
+    os.makedirs(directory, exist_ok=True)
```

Comment:
I'm concerned by this note in the doc for `makedirs`:
>Note `makedirs()` will become confused if the path elements to create include pardir (eg. “..” on UNIX systems). 

I could see someone wanting to store their stan files in `../models/` for example